cluster-autoscaler/cloudprovider/gce/autoscaling_gce_client.go

```diff
@@ -166,22 +167,67 @@ func (client *autoscalingGceClientV1) DeleteInstances(migRef GceRef, instances [
 	return client.waitForOp(op, migRef.Project, migRef.Zone)
 }
 
-func (client *autoscalingGceClientV1) FetchMigInstances(migRef GceRef) ([]GceRef, error) {
-	registerRequest("instance_group_managers", "list_managed_instances")
+func (client *autoscalingGceClientV1) FetchMigInstances(migRef GceRef) ([]InstanceInfo, error) {
 	instances, err := client.gceService.InstanceGroupManagers.ListManagedInstances(migRef.Project, migRef.Zone, migRef.Name).Do()
 	if err != nil {
 		klog.V(4).Infof("Failed MIG info request for %s %s %s: %v", migRef.Project, migRef.Zone, migRef.Name, err)
 		return nil, err
 	}
-	refs := []GceRef{}
-	for _, i := range instances.ManagedInstances {
-		ref, err := ParseInstanceUrlRef(i.Instance)
+	infos := []InstanceInfo{}
+	for _, instance := range instances.ManagedInstances {
+		ref, err := ParseInstanceUrlRef(instance.Instance)
 		if err != nil {
 			return nil, err
 		}
-		refs = append(refs, ref)
+
+		info := InstanceInfo{}
+		info.GceRef = ref
+
+		switch instance.CurrentAction {
+		case "CREATING":
+			fallthrough
+		case "RECREATING":
+			fallthrough
+		case "CREATING_WITHOUT_RETRIES":
+			info.Status = BEING_CREATED
+		case "ABANDONING":
+			fallthrough
+		case "DELETING":
+			info.Status = BEING_DELETED
+		default:
+			info.Status = RUNNING
+		}
+
+		info.ErrorClass = NONE
+		if info.Status == BEING_CREATED && instance.LastAttempt != nil && instance.LastAttempt.Errors != nil && instance.LastAttempt.Errors.Errors != nil {
+			for _, instanceError := range instance.LastAttempt.Errors.Errors {
+				if isStockoutErrorCode(instanceError.Code) {
+					info.ErrorClass = STOCKOUT
+				} else if isQuotaExceededErrorCoce(instanceError.Code) {
+					info.ErrorClass = QUOTA_EXCEEDED
+				} else if info.ErrorClass == NONE {
+					info.ErrorClass = OTHER
+				}
+				if instanceError.Message != "" {
+					if info.ErrorMessage != "" {
+						info.ErrorMessage += "; "
+					}
+					info.ErrorMessage += instanceError.Message
```

Comment:
It's probably premature optimization, but it hurts me to see string being build this way. 

strings.Builder?

Reply:
Out of curiosity, is gathering a list and calling strings.Join significantly terribly inefficient? We seem to use that a lot

Reply:
Nah, it's ok. I just didn't notice your comment as it was already marked as resolved. I consider both ways equally good, I just thought of Builder first.

Python developer in me feels a bit ashamed about that :s